test/test_nn.py

```diff
@@ -7844,7 +7947,15 @@ def perm_fn(x):
                                               )).to(device)
             self.assertEqual(tuple(result.shape), tuple(ref_output.shape))
             torch.testing.assert_close(result, ref_output, rtol=1e-7, atol=1e-5)
-
+        for batch_first in (True, False):
+            for training in (True, False):
+                # Fast path requires inference mode.
+                if training:
+                    cm = contextlib.nullcontext()
+                else:
+                    cm = torch.inference_mode()
```

Comment:
no_grad

note that inference mode sets no grad internally. So no_grad is stricter